### DISCUSSION TITLE
Optimize me #1, aka performance of wheelchairs

### DISCUSSION THREAD

**Discussion by SomeRandomNoobKekeke** [FakeFishGames/Barotrauma#15204] [Ideas and suggestions]
I mean update time of Controller ItemComponent
It's not a lot but kinda unexpected, at least for me

The most cpu consuming part is this condition
https://github.com/FakeFishGames/Barotrauma/blob/7d5b7a310a52372dac0a8bac90fd2a6a2ab342dc/Barotrauma/BarotraumaShared/SharedSource/Items/Components/Machines/Controller.cs#L245
From here
https://github.com/FakeFishGames/Barotrauma/blob/7d5b7a310a52372dac0a8bac90fd2a6a2ab342dc/Barotrauma/BarotraumaShared/SharedSource/Items/Components/Machines/Controller.cs#L240-L257

Actual CancelUsing happens in single cases, but conditionals are checked every frame

![20241120040020_1](https://github.com/user-attachments/assets/2ad64240-5a53-4b4f-b640-68487d946186)
![20241120040026_1](https://github.com/user-attachments/assets/4b5b8caa-a7b1-45d9-adf1-7958b34033ac)
![20241120040038_1](https://github.com/user-attachments/assets/f2b666b5-b200-4e61-9921-8d2c68eea6b5)

I mean something like this 
![Screenshot_7](https://github.com/user-attachments/assets/f4b182cb-b515-4a1e-a003-7b3314d59f00)

Also by "ResetPullJoints" i mean this
https://github.com/FakeFishGames/Barotrauma/blob/7d5b7a310a52372dac0a8bac90fd2a6a2ab342dc/Barotrauma/BarotraumaShared/SharedSource/Items/Components/Machines/Controller.cs#L327-L336

And "set limb anchors" is this
https://github.com/FakeFishGames/Barotrauma/blob/7d5b7a310a52372dac0a8bac90fd2a6a2ab342dc/Barotrauma/BarotraumaShared/SharedSource/Items/Components/Machines/Controller.cs#L338-L368

**Comment by Regalis11** [comment]
Thank you for the report! This does seem needlessly heavy and relatively easy to optimize. It probably doesn't make that much difference in practice though, even with a dozen characters on periscopes the cost seems to be in the ballpark of ~1% CPU time.

I don't entirely understand all your comments here though: How would splitting the conditions to multiple variables help? As far as I can see, that would actually make the performance worse, because every condition would need to be checked, which won't happen with `||` because the operators are short-circuiting in C#.

What do you mean by "actual CancelUsing happens in single cases, but conditionals are checked every frame"? Are you implying that these checks could be skipped if `CancelUsing` doesn't need to be called? That's not the case - the result of these checks is that the periscope stops updating and the rest of the method isn't executed if there's no user or if the user can no longer access the periscope.

**Reply by SomeRandomNoobKekeke** [reply]
> Thank you for the report!

Not exactly a report, just an idea of what else could you optimize in spare time in the bright future

> seems to be in the ballpark of ~1% CPU time

Yes, especially on your machine, but it's the laggies item component, ic is the laggies part of item -> MapEntity -> Update
So it's the laggies thing in update, i would start with those
You need just 3 people on periscopes / in chairs to outlag navterminal

![20241120142201_1](https://github.com/user-attachments/assets/421d7f8b-1b1a-4ae3-a424-1c5a70608320)

> How would splitting the conditions to multiple variables help?

No, i didn't mean "split the conditions", i was just showing what those conditions on screenshots ~mean in code

> What do you mean by "actual CancelUsing happens in single cases

Isn't it the thing when you're forcedly detached from periscope by e.g. water current / monsters?
So 99% of time user continue to use it, but those "CancelUsing" conditions are checked every frame, and they are not short-circuited
I mean, just look at this, do you really need to check all of that every frame for every wheelchair?
https://github.com/FakeFishGames/Barotrauma/blob/7d5b7a310a52372dac0a8bac90fd2a6a2ab342dc/Barotrauma/BarotraumaShared/SharedSource/Characters/Character.cs#L2869-L3087

**Reply by Regalis11** [reply]
> Yes, especially on your machine, but it's the laggies item component, ic is the laggies part of item -> MapEntity -> Update
> So it's the laggies thing in update, i would start with those
> You need just 3 people on periscopes / in chairs to outlag navterminal

I think you might be misinterpreting the readings somewhat. They mean that those items are taking the most time to update, but the time they're taking is still very small (fractions of a millisecond even with a ton of characters on the periscopes). So it's debatable if it even makes sense to invest time in optimizing this rather than focusing on more impactful optimizations.

**Reply by SomeRandomNoobKekeke** [reply]
Indeed debatable, but i think there's no more impactful optimizations left
It's just a mess of small unoptimized things
![BaroModder (small)](https://github.com/user-attachments/assets/2333a16e-50ad-4ad3-b644-f5cb8a4e3a54)

**Comment by Regalis11** [comment]
We went ahead and optimized this in our private repo, changes coming up in the next unstable build

**Reply by SomeRandomNoobKekeke** [reply]
![pat pat](https://github.com/user-attachments/assets/41eab944-b9ff-489b-99e6-0ec34d1ffcfa)
